### PR TITLE
EstimateCost API

### PR DESCRIPTION
**Pull request by jpbetz**
Introduce a cost estimator that walks the parsed AST and calls a provided Estimator to estimate the lengths of variable length input data (strings, bytes, lists, maps) and the cost of extension function calls.

### CODE REVIEW

File: checker/cost.go [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+	EstimateSize(element Expr) *int64
+	// EstimateCallCost returns the estimated cost of an invocation, or nil if
+	// the estimator has no estimate to provide. EstimateCallCost is only called for functions
+	// that CEL does not already know the cost; EstimateCallCost is not called for standard operations
+	// or functions.
+	EstimateCallCost(name string, target *Expr, args []Expr) *CostEstimate
+}
+
+// Expr represents an AST node for the purpose of cost estimations.
+type Expr interface {
```
Comment (TristonianJones) [inline checker/cost.go]: Consider renaming this to `AstNode` as there are quite a few other `Expr` concepts.

File: checker/cost.go [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+	"github.com/google/cel-go/common/containers"
+	"github.com/google/cel-go/common/types/ref"
+)
+
+// CostEstimator estimates the sizes of variable length input data and the costs of functions.
+type CostEstimator interface {
+	// EstimateSize returns the worst case size possible for the given Expr, or nil if
+	// the estimator has no estimate to provide. EstimateSize is only called for Exprs where
+	// CEL does not know the size; EstimateSize is not called for literals defined inline in CEL.
+	EstimateSize(element Expr) *int64
```
Comment (TristonianJones) [inline checker/cost.go]: Consider returning a `*SizeEstimate` here. though the size feels more like estimating the bounds. Perhaps, `EstimateBounds`?

Another option is to just reuse `CostEstimate` as the result from both calls, though the name might need to change if this were the case.

File: checker/cost.go [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+	EstimateCallCost(name string, target *Expr, args []Expr) *CostEstimate
+}
+
+// Expr represents an AST node for the purpose of cost estimations.
+type Expr interface {
+	// Type returns the deduced type of the Expr.
+	Type() *exprpb.Type
+	// LiteralSize returns the size of the Expr if it is a literal defined inline in CEL, or
+	// nil if the Exprs size is not statically known.
+	LiteralSize() *int64
```
Comment (TristonianJones) [inline checker/cost.go]: Usually when I think of literals, I think of constants, but a list or struct literal would have very clear sizes. Perhaps, just calling this method `FixedSize()` and returning a pointer to an int would be sufficient.

File: checker/cost.go [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+		switch ck := ek.ConstExpr.ConstantKind.(type) {
+		case *exprpb.Constant_StringValue:
+			v = int64(len(ck.StringValue))
+		case *exprpb.Constant_BytesValue:
+			v = int64(len(ck.BytesValue))
+		default:
+			return nil
+		}
+	case *exprpb.Expr_ListExpr:
+		v = int64(len(ek.ListExpr.Elements))
```
Comment (TristonianJones) [inline checker/cost.go]: It's worth a comment about how list size is used here. 

The list _could_ be the size of whatever elements it contains. For example a set membership test of `string in [a, b, c]` it might be relevant to know if either `a`, `b`, or `c` were quite large. It's very probable that you've accounted for this elsewhere, but in the event you haven't, then it's worth spending more time to hammer out.

File: checker/cost.go [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+	case *exprpb.Expr_StructExpr:
+		if ek.StructExpr.MessageName == "" {
+			v = int64(len(ek.StructExpr.Entries))
```
Comment (TristonianJones) [inline checker/cost.go]: The entry set may be composed of expression keys and expression values. Do you want to just consider the entries as size or the sum of the sizes of nested entry key and value expressions?

File: checker/cost.go [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+)
+
+// CostEstimate represents an estimated cost range.
+type CostEstimate struct {
+	Min, Max int64
+}
+
+// Add adds the costs and returns the sum.
+func (ce CostEstimate) Add(rhs CostEstimate) CostEstimate {
+	return CostEstimate{
```
Comment (TristonianJones) [inline checker/cost.go:149]: Any thoughts on whether to use a pointer instead? It's fine to use a value, just curious though since in the `Expr` interface it's not clear whether the pointer is being used to make the return "optional", or if the pointer is used to indicate that the value is longer lived somewhow.

File: checker/cost.go [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+
+	// undeclared reference
+	c.setType(e, decls.Error)
+	return CostEstimate{}
+}
+
+func (c *coster) costSelect(e *exprpb.Expr) CostEstimate {
+	sel := e.GetSelectExpr()
+	var sum CostEstimate
+	// Before traversing down the tree, try to interpret as qualified name.
```
Comment (TristonianJones) [inline checker/cost.go]: I would refrain from copying the checker code here. All of the information you want should be in the `references` and `types` derived from the `CheckedExpr.reference_map` and `CheckedExpr.type_map` respectively.

File: checker/cost.go [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+	case *exprpb.Expr_ComprehensionExpr:
+		cost = c.costComprehension(e)
+	default:
+		return CostEstimate{}
+	}
+	return cost
+}
+
+func (c *coster) costIdent(e *exprpb.Expr) CostEstimate {
+	identExpr := e.GetIdentExpr()
```
Comment (TristonianJones) [inline checker/cost.go:259]: Ident would be the top-level variable. Usually this will be an `identCost`. There should be no need to iterate through the identifiers in the environment since the type-checker will have already done this for you and rewritten the expression to ensure that only resolved identifiers appear in `IdentExpr` nodes.

File: checker/cost.go [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+	// undeclared reference
+	c.setType(e, decls.Error)
+	return CostEstimate{}
+}
+
+func (c *coster) costSelect(e *exprpb.Expr) CostEstimate {
+	sel := e.GetSelectExpr()
+	var sum CostEstimate
+	// Before traversing down the tree, try to interpret as qualified name.
+	qname, found := containers.ToQualifiedName(e)
```
Comment (TristonianJones) [inline checker/cost.go]: Skip the qualified name resolution. It will have already been done for you.

File: checker/cost.go [hunk 10]
```diff
@@ -0,0 +1,536 @@
...
+		if isDynOrError(targetType) {
+			resultType = decls.Dyn
+		} else {
+			// does not support field selection
+		}
+	}
+	if sel.TestOnly {
+		resultType = decls.Bool
+	}
+	c.setType(e, resultType)
```
Comment (TristonianJones) [inline checker/cost.go]: I would just make the `types` immutable (skip setting the type or computing the result type) since the information should already be filled in by the type-checker.

File: checker/cost.go [hunk 11]
```diff
@@ -0,0 +1,536 @@
...
+		}
+	}
+
+	// Interpret as field selection, first traversing down the operand.
+	sum = sum.Add(c.cost(sel.Operand))
+	targetType := c.getType(sel.Operand)
+	// Assume error type by default as most types do not support field selection.
+	resultType := decls.Error
+	switch kindOf(targetType) {
+	case kindMap:
```
Comment (TristonianJones) [inline checker/cost.go]: I would combine the cases here:

```
switch kindOf(targetType) {
  case kindMap, kindObject, kindTypeParam:
    sum = sum.Add(selectCost)
}

File: checker/cost.go [hunk 12]
```diff
@@ -0,0 +1,536 @@
...
+		// Maps yield their value type as the selection result type.
+		mapType := targetType.GetMapType()
+		resultType = mapType.ValueType
+		sum.Add(selectCost)
```
Comment (TristonianJones) [inline checker/cost.go]: Don't forget to assign to sum or return here and below.

File: checker/cost.go [hunk 13]
```diff
@@ -0,0 +1,536 @@
...
+			call.Function = fn.GetName()
+			return sum
+		}
+	}
+
+	// Regular instance call.
+	var targetType Expr
+	if call.Target != nil {
+		sum = sum.Add(c.cost(call.Target))
+		targetType = expr{t: c.types[call.Target.Id], expr: call.Target}
```
Comment (TristonianJones) [inline checker/cost.go]: Consider using `expr{t: c.getType(call.GetTarget()) ...}`

File: checker/cost.go [hunk 14]
```diff
@@ -0,0 +1,536 @@
...
+	}
+	fn := c.env.LookupFunction(fnName)
+	// Function found, attempt overload resolution.
+	if fn != nil {
+		sum = sum.Add(c.functionCost(call.Function, &targetType, argTypes))
```
Comment (TristonianJones) [inline checker/cost.go]: The `reference_map` should indicate the specific overload id for the function being used. It's probably worthwhile to create a map of function overload id to `costFactor`

File: checker/cost.go [hunk 15]
```diff
@@ -0,0 +1,536 @@
...
+	}
+	// Function name not declared.
+	return CostEstimate{}
+}
+
+func (c *coster) costCreateList(e *exprpb.Expr) CostEstimate {
+	create := e.GetListExpr()
+	var elemType *exprpb.Type
+	var sum CostEstimate
+	for _, e := range create.Elements {
```
Comment (TristonianJones) [inline checker/cost.go]: I feel like you need to either compute the cost in the aggregate or per-element. The per-element approach is likely to be more accurate in many cases as the use case for scaling expression cost by list or map size is usually when using a list or map variable whose size cannot be computed statically.

File: checker/cost.go (hunk 5 above)
Comment (jpbetz) [reply checker/cost.go]: I've clarified that size is the same as what CEL size() computes (lengths of strings/bytes, # of list items, # of map entries).

File: checker/cost.go (hunk 6 above)
Comment (jpbetz) [reply checker/cost.go:149]: A non-pointer seemed appropriate. It's just a composite scalar with operations that don't mutate state.

File: checker/cost.go (hunk 7 above)
Comment (jpbetz) [reply checker/cost.go]: Ahh, this is what I needed. The duplication was really bothering me. This trims the code up substantially.

File: checker/cost.go [hunk 16]
```diff
@@ -0,0 +1,645 @@
...
+func (ce CostEstimate) MultiplyByCostFactor(costPerUnit float64) CostEstimate {
+	return CostEstimate{
+		multiplyByCostFactor(ce.Min, costPerUnit),
+		multiplyByCostFactor(ce.Max, costPerUnit),
+	}
+}
+
+// addUint64NoOverflow adds non-negative ints. If the result is exceeds math.MaxUint64, math.MaxUint64
+// is returned.
+func addUint64NoOverflow(x, y uint64) uint64 {
```
Comment (TristonianJones) [inline checker/cost.go:157]: I'm okay with exporting the overflow.go calculations if we can just re-use them here.

File: checker/cost.go (hunk 15 above)
Comment (jpbetz) [reply checker/cost.go]: Is createList only for when a list is declared inline in CEL, e.g. `[1, 2, 3 + 5 / 2]`? If so I think the per-element approach is right.

File: checker/cost.go (hunk 10 above)
Comment (jpbetz) [reply checker/cost.go]: I was _almost_ able to make this work. I have one remaining setType that in costIdent that ensures that the type of identifiers is the exact same type reference that was provided by the typeProvider. I haven't isolated why this is needed yet. It might be better to look at this one together offline as it's a bit tough to explain.

File: checker/cost.go (hunk 4 above)
Comment (jpbetz) [reply checker/cost.go]: I've clarified that size is the same as what CEL `size()` computes (lengths of strings/bytes, # of list items, # of map entries).

**Comment by jpbetz** [comment]
Feedback applied. This is ready for another pass @TristonianJones

File: cel/env.go [hunk 17]
```diff
@@ -405,6 +405,12 @@ func (e *Env) ResidualAst(a *Ast, details *EvalDetails) (*Ast, error) {
 	return checked, nil
 }
 
+// EstimateCost estimates the cost of the CEL expression using the length estimates of input data and extension functions
+// provided by estimator.
+func (e *Env) EstimateCost(ast *exprpb.CheckedExpr, estimator checker.CostEstimator) checker.CostEstimate {
```
Comment (jpbetz) [inline cel/env.go]: @TristonianJones Should we include something like this? Would you expect it to automatically make sure the checker is run?

File: checker/cost.go (hunk 8 above)
Comment (jpbetz) [reply checker/cost.go:259]: I agree. This should be removed. I hit a snag and added a TODO to try to explain. It might be worth syncing offline about this one. I'm not sure I understand the codebase well enough to resolve the problem.

File: cel/env.go (hunk 17 above)
Comment (jpbetz) [reply cel/env.go]: Nevermind, I think I figured out how to do this reasonably cleanly.

File: checker/cost.go (hunk 14 above)
Comment (jpbetz) [reply checker/cost.go]: Great, I've incorporated overload ids.

File: checker/cost.go [hunk 18]
```diff
@@ -0,0 +1,432 @@
...
+
+// SizeEstimate represents an estimated size of a variable length string, bytes, map or list.
+type SizeEstimate struct {
+	Min, Max uint64
+}
+
+// Multiply multiplies by another SizeEstimate and returns the sum.
+// If multiply would result in an uint64 overflow, the result is Maxuint64.
+func (ce SizeEstimate) Multiply(sizeEstimate SizeEstimate) CostEstimate {
+	return CostEstimate{
```
Comment (TristonianJones) [inline checker/cost.go]: Shouldn't this be a `SizeEstimate` here?

File: checker/cost.go [hunk 19]
```diff
@@ -0,0 +1,432 @@
...
+}
+
+// SizeEstimate represents an estimated size of a variable length string, bytes, map or list.
+type SizeEstimate struct {
+	Min, Max uint64
+}
+
+// Multiply multiplies by another SizeEstimate and returns the sum.
+// If multiply would result in an uint64 overflow, the result is Maxuint64.
+func (ce SizeEstimate) Multiply(sizeEstimate SizeEstimate) CostEstimate {
```
Comment (TristonianJones) [inline checker/cost.go]: Prefer `se` or `e` for the `SizeEstimate`

File: checker/cost.go [hunk 20]
```diff
@@ -0,0 +1,432 @@
...
+// Cost estimates the cost of the parsed CEL expression.
+// The parsedExpr is expected to have passed type checking. If it has not been
+// type checked, the results are undefined.
+func Cost(checker *exprpb.CheckedExpr,
+	env *Env, estimator CostEstimator) CostEstimate {
+	c := coster{
+		env:       env,
+		checker:   checker,
+		estimator: estimator,
+	}
```
Comment (TristonianJones) [inline checker/cost.go:228]: Save the `type_map` from the `CheckedExpr` here so that you can refer to the type by the `Expr.GetId()` rather than trying to compute it dynamically, e.g. 

```go
func (c *coster) getType(expr *exprpb.Expr) {
  return c.type_map[expr.GetId()]
}
```

File: checker/cost.go (hunk 8 above)
Comment (TristonianJones) [reply checker/cost.go:259]: I added a comment about how to use the `CheckedExpr.type_map` near the `coster` struct definition. Hopefully that helps with removing the need for the `c.env.IdentLookup`

File: checker/cost.go [hunk 21]
```diff
@@ -0,0 +1,432 @@
...
+		return *est
+	}
+	switch overloadId {
+	// O(n) functions
+	case overloads.StartsWithString, overloads.EndsWithString, overloads.StringToBytes, overloads.BytesToString:
+		if len(args) == 1 {
+			return c.sizeEstimate(args[0]).MultiplyByCostFactor(0.1)
+		}
+	case overloads.InList:
+		// TODO: account for optimization where lists are converted to sets?
```
Comment (TristonianJones) [inline checker/cost.go]: If the `CreateList` is composed entirely of constant values, then the optimization applies, but you'd need to get this information from the `CheckedExpr` args. It's probably fine the leave the cost computation as-is here.

File: checker/cost.go [hunk 22]
```diff
@@ -0,0 +1,432 @@
...
+		// Set the range iteration variable to type DYN as well.
+		varType = decls.Dyn
+	default:
+		// not a valid range
+		varType = decls.Error
+	}
+
+	// Create a scope for the comprehension since it has a local accumulation variable.
+	// This scope will contain the accumulation variable used to compute the result.
+	c.env = c.env.enterScope()
```
Comment (TristonianJones) [inline checker/cost.go]: With the `type_map` change, I think you can remove the dependence on the `*Env` and the scope management used below.

File: checker/cost.go (hunk 18 above)
Comment (jpbetz) [reply checker/cost.go]: Yes, good catch.

File: checker/cost.go (hunk 16 above)
Comment (jpbetz) [reply checker/cost.go:157]: I'd really like to not have to define these. They're subtly different.  Any operation that would evaluate to a value greater than MaxInt instead evaluates to MaxInt.  I've also used uint64 throughout this code since cost should be non-negative and it seemed simpler to have that enforced by the type.

I'm on the fence but currently inclined to keep these locally defined due to the differences.

File: checker/cost.go (hunk 20 above)
Comment (jpbetz) [reply checker/cost.go:228]: Okay, I've dug into this and figured out what's going wrong.

The problem I'm having is that I need to attach maxLength information to `exprpb.Type`. The way I'm doing that here is to build up a "hints" `map[exprpb.Type]CostEstimate`.  But Check substitutes the `exprpb.Type` in `CheckedExpr.TypeMap` (for good reason) so I lose the keys in the mapping.

Alternatives:
- Check returns an unsubstituted mapping (not a real option, this is unsafe)
- Check somehow returns access to its `exprpb.Type` -> `exprpb.Type` `mappings` so I can follow the substitutions
- I continue to reconstruct the `Expr`->`Type` mapping (a poor option, I agree)
- Some alternative I haven't thought of

(Note that two identical `exprpb.Type`s might have different CostEstimates, so I can't identify them structurally)

Comment (jpbetz) [reply checker/cost.go:228]: Per offline discussion, I've added field paths. They seem to work fine.

File: checker/cost.go (hunk 8 above)
Comment (jpbetz) [reply checker/cost.go:259]: https://github.com/google/cel-go/pull/483/commits/49c0f67e25e528a1d7827caf6cd00d972e98de28#r809444558

File: checker/cost.go (hunk 22 above)
Comment (jpbetz) [reply checker/cost.go]: For field paths I still need to do some scope tracking but I was able to drop all the type computations.

File: cel/cel_test.go [hunk 23]
```diff
@@ -1190,38 +1194,296 @@ func TestCustomInterpreterDecorator(t *testing.T) {
...
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.hints == nil {
+				tc.hints = map[string]int64{}
+			}
+			descriptor := new(proto3pb.TestAllTypes).ProtoReflect().Descriptor()
+			e, err := NewEnv(Declarations(tc.decls...), DeclareContextProto(descriptor))
+			if err != nil {
+				t.Fatalf("environment creation error: %s\n", err)
+			}
+			e, _ = e.Extend(CustomTypeAdapter(types.DefaultTypeAdapter))
```
Comment (TristonianJones) [inline cel/cel_test.go]: You can move up the `CustomTypeAdapter()` into the first `NewEnv()` call.

File: cel/cel_test.go [hunk 24]
```diff
@@ -1190,38 +1194,296 @@ func TestCustomInterpreterDecorator(t *testing.T) {
...
-	if min != wantedCost[0] || max != wantedCost[1] {
-		log.Fatalf("Got cost interval [%v, %v], wanted %v",
-			min, max, wantedCost)
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.hints == nil {
+				tc.hints = map[string]int64{}
+			}
+			descriptor := new(proto3pb.TestAllTypes).ProtoReflect().Descriptor()
```
Comment (TristonianJones) [inline cel/cel_test.go]: I don't think you're actually using the `TestAllTypes` proto in this test, so you can remove this line and the `DeclareContextProto()` option below.

File: cel/env.go [hunk 25]
```diff
@@ -405,6 +406,16 @@ func (e *Env) ResidualAst(a *Ast, details *EvalDetails) (*Ast, error) {
 	return checked, nil
 }
 
+// EstimateCost estimates the cost of a type checked CEL expression using the length estimates of input data and
+// extension functions provided by estimator.
+func (e *Env) EstimateCost(ast *Ast, estimator checker.CostEstimator) (checker.CostEstimate, error) {
+	if !ast.IsChecked() {
+		return checker.CostEstimate{}, fmt.Errorf("EsimateCost may only be called with a type checked Ast")
+	}
+	checked := &exprpb.CheckedExpr{Expr: ast.Expr(), ReferenceMap: ast.refMap, TypeMap: ast.typeMap, SourceInfo: ast.SourceInfo()}
```
Comment (TristonianJones) [inline cel/env.go]: Prefer `checked := AstToCheckedExpr(ast)`

File: checker/cost.go [hunk 26]
```diff
@@ -0,0 +1,430 @@
...
+	EstimateSize(element AstNode) *SizeEstimate
+	// EstimateCallCost returns the estimated cost of an invocation, or nil if
+	// the estimator has no estimate to provide.
+	EstimateCallCost(overloadId string, target *AstNode, args []AstNode) *CostEstimate
+}
+
+// AstNode represents an AST node for the purpose of cost estimations.
+type AstNode interface {
+	// Path returns a path to the AstNode. The first path element is a variable. All subsequent path elements are
+	// field selectors or '*' to indicate traversal into the elements or  of a map of list.
```
Comment (TristonianJones) [inline checker/cost.go]: nit: s/or of/of/

File: checker/cost.go [hunk 27]
```diff
@@ -0,0 +1,430 @@
...
+		return CostEstimate{}
+	}
+	return cost
+}
+
+func (c *coster) costIdent(e *exprpb.Expr) CostEstimate {
+	identExpr := e.GetIdentExpr()
+
+	// resolve the field path
+	if ident := c.env.LookupIdent(identExpr.GetName()); ident != nil {
```
Comment (TristonianJones) [inline checker/cost.go]: The lookup here indicates whether something a variable or a type identifier, but by the time an expression is checked you can be guaranteed that the identifier will be found and that both have the same `identCost`.

Adding a reference to the `c.exprPath` for a type identifier won't do any harm, so I would remove the differentiation between variables and types with the `LookupIdent` call and just use the inner if-else block to differentiate between iter and non-iter identifiers.

File: checker/cost.go [hunk 28]
```diff
@@ -0,0 +1,430 @@
...
+}
+
+func (c *coster) costCreateMap(e *exprpb.Expr) CostEstimate {
+	mapVal := e.GetStructExpr()
+	var sum CostEstimate
+	for _, ent := range mapVal.GetEntries() {
+		key := ent.GetMapKey()
+		sum = sum.Add(c.cost(key))
+
+		sum = sum.Add(c.cost(ent.Value))
```
Comment (TristonianJones) [inline checker/cost.go]: Try to consistently use the getter methods here and elsewhere on proto field accesses. I believe at some point only getters will be supported, e.g. `ent.GetValue()`

File: checker/cost.go [hunk 29]
```diff
@@ -0,0 +1,430 @@
...
+	// EstimateCallCost returns the estimated cost of an invocation, or nil if
+	// the estimator has no estimate to provide.
+	EstimateCallCost(overloadId string, target *AstNode, args []AstNode) *CostEstimate
+}
+
+// AstNode represents an AST node for the purpose of cost estimations.
+type AstNode interface {
+	// Path returns a path to the AstNode. The first path element is a variable. All subsequent path elements are
+	// field selectors or '*' to indicate traversal into the elements or  of a map of list.
+	Path() []string
```
Comment (TristonianJones) [inline checker/cost.go:43]: It's possible you'll want to consider using the `AttributePattern` in the future as it will give you features like `input.*.field` which may be desirable for handling matches against element access while still retaining fine-grained estimates for sub-element access. 

Using `AttributePattern` would require this code be moved into another package though, probably something like `cel/internal` (new).

File: checker/cost.go [hunk 30]
```diff
@@ -0,0 +1,430 @@
...
+
+func (c *coster) costComprehension(e *exprpb.Expr) CostEstimate {
+	comp := e.GetComprehensionExpr()
+	var sum CostEstimate
+	sum = sum.Add(c.cost(comp.IterRange))
+	sum = sum.Add(c.cost(comp.AccuInit))
+
+	c.env = c.env.enterScope()
+	c.env.Add(decls.NewVar(comp.AccuVar, nil)) // Don't need to track types
+	c.env = c.env.enterScope()
```
Comment (TristonianJones) [inline checker/cost.go]: I think you can remove the scope entry / exit based on the simplification suggestion related to simplifying identifier resolution.

File: checker/cost.go (hunk 26 above)
Comment (jpbetz) [reply checker/cost.go]: I switched to this scheme for field path elements (after the var name): field name, '@items', '@keys', '@values'.

File: checker/cost.go (hunk 28 above)
Comment (jpbetz) [reply checker/cost.go]: I've given it the college try :)

File: checker/cost.go (hunk 29 above)
Comment (jpbetz) [reply checker/cost.go:43]: Sounds good. I've added some test around edge cases where I need to distinguish between the map key type and value type, so I landed on using "@items", "@keys" and "@values" instead of "*" in the path.

File: checker/cost.go (hunk 30 above)
Comment (jpbetz) [reply checker/cost.go]: With scopes I lose 'items.@keys' field path for "k.contains" in this case:

```
input.all(k, input[k].all(k, true) && k.contains(k))
```

Comment (jpbetz) [reply checker/cost.go]: I'll use a really simple stack to get around this.

File: checker/cost.go (hunk 27 above)
Comment (jpbetz) [reply checker/cost.go]: SGTM. I've tried to simplify this to match. Let me know if I misunderstood.

File: cel/cel_test.go (hunk 24 above)
Comment (jpbetz) [reply cel/cel_test.go]: I'm using `TestAllTypes` to represent message types in the tests.

Comment (TristonianJones) [reply cel/cel_test.go]: Ah, right. Though I think the `DeclareContextProto` isn't doing what you expect. It should be publishing a bunch of declarations for all of the type-level fields in the proto.

File: checker/cost.go [hunk 31]
```diff
@@ -0,0 +1,462 @@
...
+func (c *coster) costCall(e *exprpb.Expr) CostEstimate {
+	call := e.GetCallExpr()
+	target := call.GetTarget()
+	args := call.GetArgs()
+
+	var sum CostEstimate
+
+	argTypes := make([]AstNode, len(args))
+	for i, arg := range args {
+		// TODO: && || short circuit, so min cost should only include 1st arg eval
```
Comment (TristonianJones) [inline checker/cost.go]: Similar for ternary. Min cost should be `cond + min(a, b)` within `<cond> ? a : b`. Probably worth an additional comment.

Comment (jpbetz) [reply checker/cost.go]: +1, added a note about comprehension short circuiting as well. The minimums are not quite right in the current implementation.

File: checker/cost.go [hunk 32]
```diff
@@ -0,0 +1,468 @@
...
+	}
+	return 0, false
+}
+
+// Cost estimates the cost of the parsed CEL expression.
+// The parsedExpr is expected to have passed type checking. If it has not been
+// type checked, the results are undefined.
+func Cost(checker *exprpb.CheckedExpr, estimator CostEstimator) CostEstimate {
+	c := coster{
+		checker:    checker,
```
Comment (TristonianJones) [inline checker/cost.go]: nit: prefer `expr` or `checkedExpr` since `checker` is the name of the package and the input variable is just data and not doing any computation.

File: checker/cost.go [hunk 33]
```diff
@@ -0,0 +1,468 @@
...
+func (vs iterRangeScopes) peek(varName string) (int64, bool) {
+	varStack := vs[varName]
+	if len(varStack) > 0 {
+		return varStack[len(varStack)-1], true
+	}
+	return 0, false
+}
+
+// Cost estimates the cost of the parsed CEL expression.
+// The parsedExpr is expected to have passed type checking. If it has not been
```
Comment (TristonianJones) [inline checker/cost.go]: You probably don't need a note on `parsedExpr` here.

File: cel/cel_test.go (hunk 24 above)
Comment (jpbetz) [reply cel/cel_test.go]: Oh, I get it. I've switched this to use `Types(...)`

File: checker/cost.go (hunk 33 above)
Comment (jpbetz) [reply checker/cost.go]: Yeah, that's stale. Removing.

**Review by TristonianJones — APPROVED** [review]
I just have a couple of last minor nits, but otherwise looks good. Thanks @jpbetz!

File: checker/cost.go (hunk 32 above)
Comment (jpbetz) [reply checker/cost.go]: checkedExpr SGTM. Updated.
